### PR TITLE
Channel participation API without system channel

### PR DESCRIPTION
**Pull request by tock-ibm**

### CODE REVIEW

**Comment by jyellick** [comment]
This generally looks good to me.  I'd like to see if any other reviews have any higher level objections before doing a thorough review though.

File: text/0000-channel-participation-api-without-system-channel.md
Comment (mastersingh24) [inline text/0000-channel-participation-api-without-system-channel.md]:
```suggestion
of a system channel, which improves the privacy and scalability of the ordering service and the network as a whole. 
```

File: text/0000-channel-participation-api-without-system-channel.md
Comment (mastersingh24) [inline text/0000-channel-participation-api-without-system-channel.md]:
```suggestion
A system channel requires that all ordering service nodes (OSNs) be aware of all channels. As Fabric deprecated the 
```

File: text/0000-channel-participation-api-without-system-channel.md
Comment (mastersingh24) [inline text/0000-channel-participation-api-without-system-channel.md]:
```suggestion
Kafka-based ordering service and moved to Raft, it is no longer 
```

No need to mention any future stuff here IMHO

File: text/0000-channel-participation-api-without-system-channel.md
Comment (mastersingh24) [inline text/0000-channel-participation-api-without-system-channel.md]:
```suggestion
channels increase. However, the system channel is an exception; decoupling application channel consenters sets as 
```

File: text/0000-channel-participation-api-without-system-channel.md
Comment (mastersingh24) [inline text/0000-channel-participation-api-without-system-channel.md]:
```suggestion
of the organization - in order to apply it to the newly created channel. This creates the convoluted situation in 
```

File: text/0000-channel-participation-api-without-system-channel.md
Comment (mastersingh24) [inline text/0000-channel-participation-api-without-system-channel.md]: These should probably be a bulleted list for better readability

File: text/0000-channel-participation-api-without-system-channel.md
Comment (mastersingh24) [inline text/0000-channel-participation-api-without-system-channel.md]:
```suggestion
We introduce a new mode of channel management that does not depend on the existence of a system channel.
```

File: text/0000-channel-participation-api-without-system-channel.md
Comment (mastersingh24) [inline text/0000-channel-participation-api-without-system-channel.md:78]: Do we really need to be fine-grained here or is a local orderer admin fine (similar to a local peer admin)?

File: text/0000-channel-participation-api-without-system-channel.md
Comment (mastersingh24) [inline text/0000-channel-participation-api-without-system-channel.md:87]: Does this have to be the case?  Seems like it should be possible to support both ... but I'll read about "migration" first

File: text/0000-channel-participation-api-without-system-channel.md
Comment (mastersingh24) [inline text/0000-channel-participation-api-without-system-channel.md]:
```
... it should be possible to start OSN processes without channels, create a system channel with the
```
This is an important point ... let's make sure it does not get lost

File: text/0000-channel-participation-api-without-system-channel.md
Comment (mastersingh24) [inline text/0000-channel-participation-api-without-system-channel.md]:
```suggestion
If the OSN is joining a new channel, the config block is the genesis block of the channel. It looks like the first block of an 
```
we don't need to say how the genesis block is created here as it's not relevant

File: text/0000-channel-participation-api-without-system-channel.md
Comment (mastersingh24) [inline text/0000-channel-participation-api-without-system-channel.md]:
```suggestion
If the OSN is in the consenters set, it will continue as **member**, first pulling blocks from other OSNs util it catches
```

File: text/0000-channel-participation-api-without-system-channel.md
Comment (mastersingh24) [inline text/0000-channel-participation-api-without-system-channel.md:158]:
```
The solution is to join the new OSN as a follower. After it catches up with the cluster the 
channel admin can update the channel config and include it in the consenters set.
```
This is actually a new feature as well, correct?

File: text/0000-channel-participation-api-without-system-channel.md
Comment (mastersingh24) [inline text/0000-channel-participation-api-without-system-channel.md]:
```suggestion
owning organization will be defined in the channel configuration, and that the signature of the new OSN on the `Deliver` 
```

File: text/0000-channel-participation-api-without-system-channel.md
Comment (mastersingh24) [inline text/0000-channel-participation-api-without-system-channel.md]:
```suggestion
request satisfies the `/Channel/Readers` policy.
```

File: text/0000-channel-participation-api-without-system-channel.md
Comment (mastersingh24) [inline text/0000-channel-participation-api-without-system-channel.md]: You should not mention `configtxgen` here ... it's just a tool.
I think the important thing to detail here is what needs to be in a channel config and the fact that there is no longer a consortium defined.  This means that you can no longer simply reference MSP content but must explicitly include it in the channel create / config transaction / genesis block.

At least I think that's the case

**Review by mastersingh24 — CHANGES_REQUESTED** [review]
Overall, I'm really happy about the general content and idea here.

**Comment by mastersingh24** [comment]
> This generally looks good to me. I'd like to see if any other reviews have any higher level objections before doing a thorough review though.

No objections from me on the overall concept.  Can you review the actual APIs in detail?

File: text/0000-channel-participation-api-without-system-channel.md
Comment (sykesm) [inline text/0000-channel-participation-api-without-system-channel.md:128]: The roles of "follower" and "member" should be more formally defined. It's referenced here with assumptions about the reader's definition of those words which may not match the intent.

I realize the term was emphasized earlier with an implied definition but I believe an explicit definition of the concept would make this more straightforward.

File: text/0000-channel-participation-api-without-system-channel.md
Comment (sykesm) [inline text/0000-channel-participation-api-without-system-channel.md]: `configtx.yaml` should not be explicitly referenced here. It's a (poor) means to an end.

**Comment by jyellick** [comment]
> > This generally looks good to me. I'd like to see if any other reviews have any higher level objections before doing a thorough review though.
> 
> No objections from me on the overall concept. Can you review the actual APIs in detail?

Perhaps I should have been a bit clearer.  Prior to publication, I reviewed this quite a bit and after a few rounds of communication was pretty happy with it.  Since not much has changed I was planning to wait to do a thorough re-review once feedback from the other reviewers was incorporated.

File: text/0000-channel-participation-api-without-system-channel.md
Comment (sykesm) [inline text/0000-channel-participation-api-without-system-channel.md:109]: It's a detail but can we add a small section on API versioning please? Resource versioning? Accept headers?

Or are we simply using REST as nomenclature instead of the target implementation pattern?

File: text/0000-channel-participation-api-without-system-channel.md
Comment (sykesm) [inline text/0000-channel-participation-api-without-system-channel.md]: There is no "operations API". There's an embedded server that this can hook into to provide the orderer API. We must keep them separate.

The implementation also does *not* belong in core/operations.

File: text/0000-channel-participation-api-without-system-channel.md
Comment (sykesm) [inline text/0000-channel-participation-api-without-system-channel.md]: `/channels` should probably be scoped to your API instead of hung off the root.

File: text/0000-channel-participation-api-without-system-channel.md
Comment (sykesm) [inline text/0000-channel-participation-api-without-system-channel.md:503]: Where is that effort taking place? I'm very interested as it has larger impacts and is explicitly an area that I have opinions on.

**Review by sykesm — DISMISSED** [review]

File: text/0000-channel-participation-api-without-system-channel.md
Comment (tock-ibm) [reply text/0000-channel-participation-api-without-system-channel.md:158]: Yes, this is a new feature.

File: text/0000-channel-participation-api-without-system-channel.md
Comment (tock-ibm) [reply text/0000-channel-participation-api-without-system-channel.md]: Fixed

File: text/0000-channel-participation-api-without-system-channel.md
Comment (tock-ibm) [reply text/0000-channel-participation-api-without-system-channel.md]: Fixed

File: text/0000-channel-participation-api-without-system-channel.md
Comment (tock-ibm) [reply text/0000-channel-participation-api-without-system-channel.md]: Fixed

File: text/0000-channel-participation-api-without-system-channel.md
Comment (tock-ibm) [reply text/0000-channel-participation-api-without-system-channel.md]: Fixed

File: text/0000-channel-participation-api-without-system-channel.md
Comment (tock-ibm) [reply text/0000-channel-participation-api-without-system-channel.md]: Fixed

File: text/0000-channel-participation-api-without-system-channel.md
Comment (tock-ibm) [reply text/0000-channel-participation-api-without-system-channel.md]: Fixed

File: text/0000-channel-participation-api-without-system-channel.md
Comment (tock-ibm) [reply text/0000-channel-participation-api-without-system-channel.md]: Fixed

File: text/0000-channel-participation-api-without-system-channel.md
Comment (tock-ibm) [reply text/0000-channel-participation-api-without-system-channel.md]: Fixed

File: text/0000-channel-participation-api-without-system-channel.md
Comment (tock-ibm) [reply text/0000-channel-participation-api-without-system-channel.md:78]: For the sake of generality, I think it is useful to formally define this role. 
Our current design choice is to have this role will use the same AA mechanism as the "operations API" admin (see Security section). However it is conceivable that these roles would be separated some day.

File: text/0000-channel-participation-api-without-system-channel.md
Comment (tock-ibm) [reply text/0000-channel-participation-api-without-system-channel.md:87]: "Mixed mode" is a problem... things can stop working in ugly ways, e.g. on-boarding. There are solutions of course, but they are complicated and costly. Therefore, for simplicity, I think we should avoid that and encourage users to use either one or the other. 

In addition, moving from system-channel mode to participation-api style is easy (delete system-channel on all OSNs). The reverse, going back to system-channel style once there are already app-channels  is complicated. Here too we choose simplicity and do not allow the reverse path. The only exception is bootstrap: starting from empty OSNs and creating the system channel.

File: text/0000-channel-participation-api-without-system-channel.md
Comment (tock-ibm) [reply text/0000-channel-participation-api-without-system-channel.md]: I split the paragraph to make this more prominent.

File: text/0000-channel-participation-api-without-system-channel.md
Comment (tock-ibm) [reply text/0000-channel-participation-api-without-system-channel.md]: Done

File: text/0000-channel-participation-api-without-system-channel.md
Comment (tock-ibm) [reply text/0000-channel-participation-api-without-system-channel.md]: Done

File: text/0000-channel-participation-api-without-system-channel.md
Comment (tock-ibm) [reply text/0000-channel-participation-api-without-system-channel.md]: Done

File: text/0000-channel-participation-api-without-system-channel.md
Comment (tock-ibm) [reply text/0000-channel-participation-api-without-system-channel.md]: I demoted the reference to configtxgen to a single example, rather then making it a prescription for generating the genesis block. I also described the changes w.r.t. a consortium & consortiums.

File: text/0000-channel-participation-api-without-system-channel.md
Comment (tock-ibm) [reply text/0000-channel-participation-api-without-system-channel.md]: Done, I demoted that to an example.

File: text/0000-channel-participation-api-without-system-channel.md
Comment (tock-ibm) [reply text/0000-channel-participation-api-without-system-channel.md:109]: I was planing (hoping...) to go into the fine-grain details when I write the Jira item dedicated to the REST API. Here it is meant to be just a sketch, in order not to confuse the reader with too many details.

File: text/0000-channel-participation-api-without-system-channel.md
Comment (tock-ibm) [reply text/0000-channel-participation-api-without-system-channel.md:503]: Gari is doing some exploratory work on that, I believe. 
@mastersingh24

File: text/0000-channel-participation-api-without-system-channel.md
Comment (tock-ibm) [reply text/0000-channel-participation-api-without-system-channel.md:333]: You mean something like this?

`channel-participation/channels`

File: text/0000-channel-participation-api-without-system-channel.md
Comment (tock-ibm) [reply text/0000-channel-participation-api-without-system-channel.md]: Most of the implementation will be under orderer/... 
Some tiny bits that hook into the HTTP server might reside elsewhere (not sure where though).

File: text/0000-channel-participation-api-without-system-channel.md
Comment (tock-ibm) [reply text/0000-channel-participation-api-without-system-channel.md:128]: Agreed. I will improve the presentation.

File: text/0000-channel-participation-api-without-system-channel.md
Comment (tock-ibm) [reply text/0000-channel-participation-api-without-system-channel.md]: Done

File: text/0000-channel-participation-api-without-system-channel.md
Comment (yeasy) [inline text/0000-channel-participation-api-without-system-channel.md:99]: Any specific reason to use REST rather than the existing gRPC style?

Comment (tock-ibm) [reply text/0000-channel-participation-api-without-system-channel.md:99]: We want to make it easier for cloud operators to manage the OSN channel participation. This way they won't need additional tools that implement gRPC, and can use a browser, for example.

File: text/0000-channel-participation-api-without-system-channel.md
Comment (jyellick) [inline text/0000-channel-participation-api-without-system-channel.md]: Very minor typo -- "util" instead of "until"

File: text/0000-channel-participation-api-without-system-channel.md
Comment (jyellick) [inline text/0000-channel-participation-api-without-system-channel.md:366]: This follows the existing patterns of the orderer codebase, but we have been moving towards placing new code not to be consumed by external parties as a library (though this is essentially all of Fabric) into `./internal`.  I'd welcome opinions from other maintainers, but either way, I don't see the location of this package as being significant enough to hold up the RFC.

File: text/0000-channel-participation-api-without-system-channel.md
Comment (jyellick) [inline text/0000-channel-participation-api-without-system-channel.md:369]: Being the primary author of much of the multichannel code, I would not get too hung up on preserving our existing channel startup and management structure.  If this code needs to be gutted and replaced, all the better.  Regardless, this seems like an implementation detail outside the required scope of the RFC, so not worth holding anything up over.

File: text/0000-channel-participation-api-without-system-channel.md
Comment (jyellick) [inline text/0000-channel-participation-api-without-system-channel.md:372]: This seems like it might be a bit racy -- we could crash after creating the ledger folder but before writing the block.  Perhaps it would be better to reverse the order and locate the block outside the ledger dir so that we could resume creating the ledger dir if interrupted?

File: text/0000-channel-participation-api-without-system-channel.md
Comment (jyellick) [inline text/0000-channel-participation-api-without-system-channel.md]: I was initially thrown for a loop here, but I think you mean the component which implements the following/joining go routines, not the the participation service which still allows for the deletion of the system channel and listing of channels.  Could be clarified.

**Review by jyellick — APPROVED** [review]
As far as I'm concerned, this is good to go.  I've added just a few comments, but I don't think it's anything critical mostly typos, and some implementation details.

File: text/0000-channel-participation-api-without-system-channel.md
Comment (mastersingh24) [reply text/0000-channel-participation-api-without-system-channel.md:99]: I'm a bit torn on this one ... I tend to think we should probably stick with gRPC here.  But I will not hold things up ... the APIs generally look ok and it's not hard to translate these to protobuf defs.

**Review by mastersingh24 — APPROVED** [review]

File: text/0000-channel-participation-api-without-system-channel.md
Comment (tock-ibm) [reply text/0000-channel-participation-api-without-system-channel.md:372]: You are correct, this needs to be done atomically. One way to do this is to create a temporary folder with all the resources and then rename it into place. Renaming is atomic so if we fail before the rename nothing happens.

File: text/0000-channel-participation-api-without-system-channel.md
Comment (tock-ibm) [reply text/0000-channel-participation-api-without-system-channel.md]: I clarified the text.

File: text/0000-channel-participation-api-without-system-channel.md
Comment (tock-ibm) [reply text/0000-channel-participation-api-without-system-channel.md]: Fixed.

File: text/0000-channel-participation-api-without-system-channel.md
Comment (tock-ibm) [reply text/0000-channel-participation-api-without-system-channel.md:369]: OK, I'll take the liberty to streamline that code if necessary.

**Review by jyellick — APPROVED** [review]

**Comment by denyeart** [comment]
No concerns raised during Final Comment Period, let's go ahead and merge.

**Comment by SamyDgz** [comment]
@SamyDgz

**Pull request merged**
